### PR TITLE
consensus: fix integer overflow in ValidateCollateralRatio()

### PR DESCRIPTION
**Pull request by gto90**
## Motivation

Fixes DGB-SEC-001 from PR #367 security audit. `ValidateCollateralRatio()` performs `collateralAmount * oraclePrice` which overflows `int64_t` for large but legitimate values (e.g. 1M DGB at $10/DGB = 10^20 > INT64_MAX 9.2x10^18). The same overflow exists for `ddAmount * requiredRatio`.

## Description

**`src/consensus/digidollar_transaction_validation.cpp`:**
- Added overflow-safe multiplication using the check-before-multiply pattern (consistent with `dca.cpp:60-70`)
- Before multiplying, checks if the value exceeds `INT64_MAX / multiplier`
- Falls back to divide-first-then-multiply when overflow would occur

**`src/digidollar/validation.cpp`:**
- Same overflow protection for `dgbLocked * ctx.oraclePriceMicroUSD`
- Added `ddMinted > 0` guard to prevent division by zero in ratio calculation

**`src/test/digidollar_transaction_tests.cpp`:**
- Added `test_collateral_ratio_overflow_protection` test case
- Tests normal values, large collateral (1M DGB), MAX_MONEY collateral, large DD amounts, and negative/zero edge cases

## Testing

- All 36 DigiDollar transaction tests pass
- Full test suite: 2 pre-existing failures in `i2p_tests` (unrelated)
- All 3 modified files compile cleanly (only pre-existing warnings)

Refs: PR #367 (DGB-SEC-001)

### CODE REVIEW

File: src/digidollar/validation.cpp [hunk 1]
```diff
@@ -407,9 +408,16 @@ bool ValidateCollateralRatio(CAmount dgbLocked, CAmount ddMinted,
...
+    CAmount dgbValueMicroUSD;
+    const CAmount maxSafeDGB = std::numeric_limits<CAmount>::max() / ctx.oraclePriceMicroUSD;
+    if (dgbLocked > maxSafeDGB) {
+        dgbValueMicroUSD = (dgbLocked / COIN) * ctx.oraclePriceMicroUSD;
+    } else {
+        dgbValueMicroUSD = (dgbLocked * ctx.oraclePriceMicroUSD) / COIN;
+    }
     CAmount dgbValueInCents = dgbValueMicroUSD / 10000;  // Convert micro-USD to cents
-    int actualRatio = (dgbValueInCents * 100) / ddMinted;
+    int actualRatio = (ddMinted > 0) ? (dgbValueInCents * 100) / ddMinted : 0;
```
[the remark this answers is not in the document]
Comment (gto90) [reply src/digidollar/validation.cpp]: Fixed in 6964d42. Added overflow protection for `dgbValueInCents * 100` using the same divide-first pattern: checks `dgbValueInCents > INT64_MAX / 100` before multiplying, and falls back to dividing `ddMinted` by 100 first to avoid overflow.

File: src/test/digidollar_transaction_tests.cpp [hunk 2]
```diff
@@ -934,4 +934,34 @@ BOOST_FIXTURE_TEST_CASE(test_dd_opcode_validation, DigiDollarTransactionTestFixt
...
+    // 10^14 * 10^6 = 10^20 — exceeds INT64_MAX without overflow protection
+    BOOST_CHECK_NO_THROW(ValidateCollateralRatio(1000, largeCollateral, MAX_PRICE, 200));
+
+    // MAX_MONEY collateral — must not crash
+    CAmount maxCollateral = MAX_MONEY; // 21 billion DGB (21000000000 * COIN)
+    BOOST_CHECK_NO_THROW(ValidateCollateralRatio(1000, maxCollateral, MAX_PRICE, 200));
+
+    // Large ddAmount that could overflow ddAmount * requiredRatio
+    CAmount largeDDAmount = std::numeric_limits<CAmount>::max() / 100; // Near max
+    BOOST_CHECK_NO_THROW(ValidateCollateralRatio(largeDDAmount, 100 * COIN, 5000, 200));
```
[the remark this answers is not in the document]
Comment (gto90) [reply src/test/digidollar_transaction_tests.cpp]: Fixed in 6964d42. Replaced all `BOOST_CHECK_NO_THROW` calls with explicit return value assertions: `BOOST_CHECK()` for cases that should pass (large collateral easily satisfies the ratio) and `BOOST_CHECK(!)` for cases that should fail (near-max ddAmount with small collateral).

File: src/test/digidollar_transaction_tests.cpp [hunk 3]
```diff
@@ -934,4 +934,34 @@ BOOST_FIXTURE_TEST_CASE(test_dd_opcode_validation, DigiDollarTransactionTestFixt
...
+
+    // Edge case: negative and zero inputs still rejected
+    BOOST_CHECK(!ValidateCollateralRatio(-1, 100 * COIN, 5000, 200));
+    BOOST_CHECK(!ValidateCollateralRatio(1000, -1, 5000, 200));
+    BOOST_CHECK(!ValidateCollateralRatio(1000, 100 * COIN, -1, 200));
+}
```
[the remark this answers is not in the document]
Comment (gto90) [reply src/test/digidollar_transaction_tests.cpp:967]: Fixed in 6964d42. Added new `test_collateral_ratio_overflow_precision` test case that exercises values just below and above `maxSafeCollateral` (the overflow threshold). Verifies both the multiply-first and divide-first paths produce consistent results, and that a failing case is rejected regardless of which path is taken.

**Comment by DigiSwarm** [comment]
**Closing as duplicate — already fixed with a better approach on `feature/digidollar-v1`.**

### Why this is superseded:
This PR uses a "divide first to avoid overflow" pattern in `ValidateCollateralRatio()` and related functions. Our branch already uses `__int128` arithmetic for all collateral calculations (commits `090ac9b1fd`, `8a169127be`, `8de117954a`), which is strictly better:

- **No precision loss** — `__int128` gives exact results vs divide-first which loses up to 999 satoshis
- **Simpler code** — one cast vs conditional branching
- **Applied everywhere** — consensus validation, TxBuilder, RPC calculations, all using the same pattern

The vulnerability was real (SEC-001) and the analysis was correct — we just fixed it differently during the overnight red team session. Thanks @gto90 for identifying the overflow vectors.
